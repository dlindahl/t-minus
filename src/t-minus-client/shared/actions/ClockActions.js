import ClockActions from '../actionTypes/ClockActionTypes';
import emptyObj from 'empty/object';
import formatDuration from '../util/formatDuration';
import partial from 'lodash/partial';
import present from 'present';

export const IntervalValue = 50;
let interval = null;

function calculateDuration(mode, elapsedTime, maxDuration) {
  switch(mode) {
    case ClockActions.CLOCK_MODE_TIMER:
      const timeRemaining = maxDuration - elapsedTime;
      const percentComplete = Math.min(elapsedTime / maxDuration, 1);
      return [formatDuration(timeRemaining), timeRemaining, percentComplete];
    default:
      return [formatDuration(elapsedTime), null, 1]
  }
}

function determineTimerElapsed(mode, elapsedTime, maxDuration) {
  switch(mode) {
    case ClockActions.CLOCK_MODE_TIMER:
      return elapsedTime > maxDuration;
    default:
      return false;
  }
}

export function clockTick(dispatch, getState) {
  const { clock } = getState();
  if(!clock.running) {
    return;
  }
  const currentTick = present();
  const delta = currentTick - clock.lastTick;
  const elapsedTime = clock.elapsedTime + delta;
  const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, elapsedTime, clock.timerValue);
  const hasTimerElapsed = determineTimerElapsed(clock.mode, elapsedTime, clock.timerValue);
  dispatch({
    type: ClockActions.CLOCK_TICK,
    payload: {
      currentTick,
      delta,
      duration,
      elapsedTime,
      hasTimerElapsed,
      mode: clock.mode,
      percentComplete,
      timeRemaining
    }
  });
}

export function pauseClock() {
  return {
    type: ClockActions.CLOCK_PAUSED,
    payload: emptyObj
  };
}

export function resetClock() {
  return (dispatch, getState) => {
    const { clock } = getState();
    clearInterval(interval);
    interval = null;
    const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, 0, clock.timerValue);
    return dispatch({
      type: ClockActions.CLOCK_RESET,
      payload: {
        duration,
        timeRemaining,
        timerValue: clock.timerValue,
        percentComplete
      }
    });
  }
}

export function startClock() {
  return (dispatch, getState) => {
    const { clock } = getState();
    const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, clock.elapsedTime, clock.timerDuration);
    dispatch({
      type: ClockActions.CLOCK_STARTED,
      payload: {
        duration,
        percentComplete,
        startedAt: present(),
        timeRemaining
      }
    });
    interval = setInterval(partial(clockTick, dispatch, getState), IntervalValue);
  };
}

export function changeTimerValue(timerValue) {
  return (dispatch, getState) => {
    const { clock } = getState();
    const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, 0, timerValue);
    return dispatch({
      type: ClockActions.CLOCK_TIMER_CHANGED,
      payload: { duration, percentComplete, timeRemaining, timerValue }
    });
  }
}

export function stopwatchMode() {
  const mode = ClockActions.CLOCK_MODE_STOPWATCH;
  const [duration, timeRemaining, percentComplete] = calculateDuration(mode, 0);
  return {
    type: ClockActions.CLOCK_MODE_CHANGED,
    payload: {
      duration,
      mode,
      percentComplete,
      timeRemaining
    }
  }
}

export function timerMode() {
  return (dispatch, getState) => {
    const { clock } = getState();
    const mode = ClockActions.CLOCK_MODE_TIMER;
    const [duration, timeRemaining, percentComplete] = calculateDuration(mode, 0, clock.timerValue);
    return dispatch({
      type: ClockActions.CLOCK_MODE_CHANGED,
      payload: {
        duration,
        mode,
        percentComplete,
        timeRemaining
      }
    });
  }
}

export function toggleClock() {
  return (dispatch, getState) => {
    const clock = getState().clock
    if(clock.startedAt == null) {
      startClock()(dispatch, getState);
    } else if(clock.running) {
      dispatch(pauseClock());
    } else {
      dispatch(unpauseClock());
    }
  };
}

export function unpauseClock() {
  return {
    type: ClockActions.CLOCK_UNPAUSED,
    payload: {}
  };
}
