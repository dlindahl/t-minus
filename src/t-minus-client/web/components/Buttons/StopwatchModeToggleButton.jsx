import AppNavButton from './AppNavButton'
import { bindActionCreators } from 'redux'
import ClockActions from '../../../shared/actionTypes/ClockActionTypes'
import { Component, PropTypes } from 'react'
import { connect } from 'react-redux'
import Schedule from 'react-icons/lib/md/Schedule'
import { stopwatchMode } from '../../../shared/actions/ClockActions'

function getState (state) {
  return {
    active: state.clock.mode === ClockActions.CLOCK_MODE_STOPWATCH
  }
}

function getActions (dispatch) {
  return bindActionCreators({ stopwatchMode }, dispatch)
}

@connect(getState, getActions)
export default class StopwatchModeToggleButton extends Component {
  constructor (props) {
    super(props)
    this.handleClick = this.handleClick.bind(this)
  }
  handleClick () {
    this.props.stopwatchMode()
  }
  render () {
    return (
      <AppNavButton active={this.props.active} onClick={this.handleClick}>
        <Schedule/>
      </AppNavButton>
    )
  }
}

StopwatchModeToggleButton.propTypes = {
  active: PropTypes.bool,
  stopwatchMode: PropTypes.func
}

StopwatchModeToggleButton.defaultProps = {
  active: false
}
