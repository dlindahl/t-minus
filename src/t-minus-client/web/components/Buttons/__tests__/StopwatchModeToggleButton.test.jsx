import ClockActions from '../../../../shared/actionTypes/ClockActionTypes'
import configureStore from 'redux-mock-store'
import emptyObj from 'empty/object'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import StopwatchModeToggleButton from '../StopwatchModeToggleButton'

const STATES = {
  default: {
    clock: emptyObj,
    display: emptyObj
  },
  stopwatchMode: {
    clock: {
      mode: ClockActions.CLOCK_MODE_STOPWATCH
    },
    display: emptyObj
  }
}
const store = configureStore()

describe('StopwatchModeToggleButton', () => {
  it('renders the component', () => {
    const component = renderer.create(
      <Provider store={store(STATES.default)}>
        <StopwatchModeToggleButton/>
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders an active component when the clock is in stopwatch mode', () => {
    const component = renderer.create(
      <Provider store={store(STATES.stopwatchMode)}>
        <StopwatchModeToggleButton/>
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('enables stopwatch mode when clicked', () => {
    const stopwatchMode = jest.fn()
    const component = renderer.create(
      <Provider store={store(STATES.default)}>
        <StopwatchModeToggleButton.WrappedComponent
          stopwatchMode={stopwatchMode}
        />
      </Provider>
    )
    const tree = component.toJSON()
    tree.props.onClick()
    expect(stopwatchMode).toHaveBeenCalled()
  })
})
