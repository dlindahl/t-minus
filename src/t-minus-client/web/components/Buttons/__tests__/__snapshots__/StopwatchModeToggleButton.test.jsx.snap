exports[`StopwatchModeToggleButton renders an active component when the clock is in stopwatch mode 1`] = `
<button
  onClick={[Function]}
  style={
    Object {
      "background": "transparent",
      "border": 0,
      "color": "var(--Concrete)",
      "cursor": "pointer",
      "fontSize": "2rem",
      "height": 50,
      "lineHeight": 1,
      "minWidth": 50,
      "padding": 0,
    }
  }>
  <svg
    fill="currentColor"
    height="1em"
    preserveAspectRatio="xMidYMid meet"
    style={
      Object {
        "color": undefined,
        "verticalAlign": "top",
      }
    }
    viewBox="0 0 40 40"
    width="1em">
    <g>
      <path
        d="m20.9 11.6v8.8l7.5 4.4-1.3 2.2-8.7-5.4v-10h2.5z m-0.9 21.8c7.3 0 13.4-6.1 13.4-13.4s-6.1-13.4-13.4-13.4-13.4 6.1-13.4 13.4 6.1 13.4 13.4 13.4z m0-30c9.2 0 16.6 7.4 16.6 16.6s-7.4 16.6-16.6 16.6-16.6-7.4-16.6-16.6 7.4-16.6 16.6-16.6z" />
    </g>
  </svg>
  <br />
  <div
    style={
      Object {
        "color": "var(--Concrete)",
        "fontFamily": "Roboto Mono, monospace",
        "fontSize": ".8rem",
        "lineHeight": 1.5,
        "textTransform": "uppercase",
      }
    }>
    Stopwatch
  </div>
</button>
`;

exports[`StopwatchModeToggleButton renders the component 1`] = `
<button
  onClick={[Function]}
  style={
    Object {
      "background": "transparent",
      "border": 0,
      "color": "var(--Silver)",
      "cursor": "pointer",
      "fontSize": "2rem",
      "height": 50,
      "lineHeight": 1,
      "minWidth": 50,
      "padding": 0,
    }
  }>
  <svg
    fill="currentColor"
    height="1em"
    preserveAspectRatio="xMidYMid meet"
    style={
      Object {
        "color": undefined,
        "verticalAlign": "top",
      }
    }
    viewBox="0 0 40 40"
    width="1em">
    <g>
      <path
        d="m20.9 11.6v8.8l7.5 4.4-1.3 2.2-8.7-5.4v-10h2.5z m-0.9 21.8c7.3 0 13.4-6.1 13.4-13.4s-6.1-13.4-13.4-13.4-13.4 6.1-13.4 13.4 6.1 13.4 13.4 13.4z m0-30c9.2 0 16.6 7.4 16.6 16.6s-7.4 16.6-16.6 16.6-16.6-7.4-16.6-16.6 7.4-16.6 16.6-16.6z" />
    </g>
  </svg>
  <br />
  <div
    style={
      Object {
        "color": "var(--Silver)",
        "fontFamily": "Roboto Mono, monospace",
        "fontSize": ".8rem",
        "lineHeight": 1.5,
        "textTransform": "uppercase",
      }
    }>
    Stopwatch
  </div>
</button>
`;
